### PR TITLE
KV mock

### PR DESCRIPTION
**Pull request by xavi-**
- Added `setGlobalRouting` method to parallel kayvee-go
- Added `mockRouting` (See README for usage example)

Context: https://clever.atlassian.net/browse/INFRA-1962

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+    // Make a query and log its length
+    let query_start = Date.now();
+    log.gauge("QueryTime", Date.now() - query_start);
+
+    // Do something async
+    setImmediate(() => {
+        // Output structured data
+        log.infoD("DataResults", {"key": "value"}); // Sends slack message
+
+        // You can use the M alias for your key value pairs
```
Comment (nathanleiby) [inline README.md]: nit: `M` alias not relevant here

File: README.md [hunk 2]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+
+```js
+// main-test.js
+const assert = require("assert");
+
+const kv = require("../kayvee-js");
+
+const main = require("./main");
+
+kv.mockRouting(kvdone => { // Don't nest kv.mockRouting calls!!
```
Comment (nathanleiby) [inline README.md]: What happens if you nest them? Can we just throw a reasonable error instead?

File: README.md (hunk 1 above)
Comment (xavi-) [inline README.md]: Reworded in ddf06f5a4b1b5adf90778db1e64cd727701256c4

File: README.md [hunk 3]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+
+const log = new kv.logger("myApp");
+
+module.exports = (cb) => {
+    // Simple debugging
+    log.debug("Service has started");
+
+    // Make a query and log its length
+    let query_start = Date.now();
+    log.gauge("QueryTime", Date.now() - query_start);
```
Comment (nathanleiby) [inline README.md]: This doesn't seem related to sending to slack-- should remove it from the example?

File: README.md [hunk 4]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+      key: [ "value" ]
+    output:
+      type: "notifications"
+      channel: "#distribution"
+      icon: ":rocket:"
+      message: "%{key}"
+      user: "Flight Tracker"
+```
+
+### Testings
```
Comment (nathanleiby) [inline README.md]: nit: testings/tests

File: README.md [hunk 5]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+    let query_start = Date.now();
+    log.gauge("QueryTime", Date.now() - query_start);
+
+    // Do something async
+    setImmediate(() => {
+        // Output structured data
+        log.infoD("DataResults", {"key": "value"}); // Sends slack message
+
+        // You can use an object to send arbitrary key value pairs
+        log.infoD("DataResults", {"shorter": "line"});
```
Comment (xavi-) [inline README.md]: Ahh, good point.  I'll point out it won't send a slack message.

File: README.md [hunk 6]
```diff
@@ -42,6 +42,75 @@ console.error(kayvee.formatLog("test_source", kayvee.INFO, "title", {"foo" : 1,
...
+        assert.ifError(err);
+
+        let ruleCounts = kvdone();
+        assert.equal(ruleCounts["key-val"], 1);
+    });
+});
+```
+
+For more information on log routing see https://clever.atlassian.net/wiki/display/ENG/Log+Routing
+
```
Comment (nathanleiby) [inline README.md]: Could we just put this in an examples folder, and then actually execute it during the test pass? 

Then we can just link to the examples folder for a working example

File: lib/logger/logger.ts [hunk 7]
```diff
@@ -161,12 +167,45 @@ class Logger {
     const data = assign({level: logLvl}, this.globals, metadata, userdata);
     if (this.logRouter != null) {
       data._kvmeta = this.logRouter.route(data);
+    } else if (globalRouter) {
```
Comment (nathanleiby) [inline lib/logger/logger.ts]: above we have a `!= null` check .. should the checks on 168 and 170 be the same?

**Comment by nathanleiby** [comment]
lgtm, just some nits

File: lib/logger/logger.ts (hunk 7 above)
Comment (xavi-) [inline lib/logger/logger.ts]: Changed in 43babe5d0daece3d29a1f9815f8a4582d915d7cf

File: README.md (hunk 2 above)
Comment (xavi-) [inline README.md]: It becomes difficult to guarantee that we've counted all the matched rules correctly.  An error is now thrown when `mockRouting` calls are nest: a7d3a0432f27129a5fb251d14ffbee09ab08e254

File: README.md (hunk 4 above)
Comment (xavi-) [inline README.md]: Fixed in 406facee4b9a3876a29f94e3a1bc8ca24a5bf92f

File: README.md (hunk 5 above)
Comment (xavi-) [inline README.md]: Updated in 406facee4b9a3876a29f94e3a1bc8ca24a5bf92f

File: README.md (hunk 3 above)
Comment (xavi-) [inline README.md]: True.  Removed in 406facee4b9a3876a29f94e3a1bc8ca24a5bf92f

File: README.md (hunk 6 above)
Comment (xavi-) [inline README.md]: I feel they're easier to read if they're show directly in the README
